src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -295,47 +310,106 @@ private static MethodInfo GetMethodInfo<T>(Expression<T> expr)
 
-                parameters[i] = new FactoryParameterContext(constructorParameter.ParameterType, hasDefaultValue, defaultValue, parameterMap[i] ?? -1);
+                if (parameterMap[i] is not null)
+                {
+                    matchedArgCount++;
+                    if (parameterMap[i] == i)
+                    {
+                        matchedArgCountWithMap++;
+                    }
+                }
             }
-            Type declaringType = constructor.DeclaringType!;
 
-            return (IServiceProvider serviceProvider, object?[]? arguments) =>
+            // Don't optimize cases with default values or arg mapping.
+            if (hasAnyDefaultValues || matchedArgCount != matchedArgCountWithMap)
             {
-                if (serviceProvider is null)
+                return InvokeCanonical();
+            }
+
+            if (matchedArgCount == 0)
+            {
+                // All injected; use a fast path.
+                Type[] types = GetParameterTypes();
+                return useFixedValues ?
+                    (serviceProvider, arguments) => ReflectionFactory_ServiceOnly_Fixed(invoker, types, declaringType, serviceProvider) :
+                    (serviceProvider, arguments) => ReflectionFactory_ServiceOnly_Span(invoker, types, declaringType, serviceProvider);
+            }
+
+            if (matchedArgCount == constructorParameters.Length)
+            {
+                // All direct with no mappings; use a fast path.
+                return (serviceProvider, arguments) => ReflectionFactory_Direct(invoker, serviceProvider, arguments);
+            }
+
+            return InvokeCanonical();
+
+            ObjectFactory InvokeCanonical()
+            {
+                FactoryParameterContext[] parameters = GetFactoryParameterContext();
+                return useFixedValues ?
+                    (serviceProvider, arguments) => ReflectionFactory_Canonical_Fixed(invoker, parameters, declaringType, serviceProvider, arguments) :
+                    (serviceProvider, arguments) => ReflectionFactory_Canonical_Span(invoker, parameters, declaringType, serviceProvider, arguments);
+            }
+
+            Type[] GetParameterTypes()
+            {
+                Type[] types = new Type[constructorParameters.Length];
+                for (int i = 0; i < constructorParameters.Length; i++)
                 {
-                    throw new ArgumentNullException(nameof(serviceProvider));
+                    types[i] = constructorParameters[i].ParameterType;
                 }
+                return types;
+            }
+#else
+            return InvokeCanonical();
+
+            ObjectFactory InvokeCanonical()
+            {
+                FactoryParameterContext[] parameters = GetFactoryParameterContext();
+                return (serviceProvider, arguments) => ReflectionFactory_Canonical(constructor, parameters, declaringType, serviceProvider, arguments);
+            }
```

Comment:
What's the value of this nested method? Why isn't this just:

```suggestion
            FactoryParameterContext[] parameters = GetFactoryParameterContext();
            return (serviceProvider, arguments) => ReflectionFactory_Canonical(constructor, parameters, declaringType, serviceProvider, arguments);
```

Reply:
This was previous attempt at shared code (shared call to InvokeCanconical but 2 implementations) but it didn't turn out that way. Removed the method.